tests/Properties.hs

```diff
@@ -60,34 +70,34 @@ appendFileReadFile = do
     contents <- OSP.readFile (baseDir </> [osp|foo|])
     "testtest" @=? contents
 
-iomodeReadFile :: IO ()
+iomodeReadFile :: Assertion
 iomodeReadFile = do
   withSystemTempDirectory "test" $ \baseDir' -> do
     baseDir <- OSP.encodeFS baseDir'
     OSP.writeFile (baseDir </> [osp|foo|]) ""
     r <- try @IOException $ OSP.withFile (baseDir </> [osp|foo|]) ReadMode $ \h -> BS.hPut h "test"
-    Left IllegalOperation
-      @=? first ioe_type r
+    IOError Nothing IllegalOperation "hPutBuf" "handle is not open for writing" Nothing Nothing
+      @==? first (\e -> e { ioe_filename = Nothing }) r
 
-iomodeWriteFile :: IO ()
+iomodeWriteFile :: Assertion
 iomodeWriteFile = do
   withSystemTempDirectory "test" $ \baseDir' -> do
     baseDir <- OSP.encodeFS baseDir'
     OSP.writeFile (baseDir </> [osp|foo|]) ""
     r <- try @IOException $ OSP.withFile (baseDir </> [osp|foo|]) WriteMode $ \h -> BS.hGetContents h
-    Left IllegalOperation
-      @=? first ioe_type r
+    IOError Nothing IllegalOperation "hGetBuf" "handle is not open for reading" Nothing Nothing
+      @==? first (\e -> e { ioe_filename = Nothing }) r
```

Comment:
We're hiding filepath here and in other places, because `BS.hGetContents` will set the file descriptor as filepath (odd).

This seems to be due to some shenanigans in base:

```hs
augmentIOError :: IOException -> String -> Handle -> IOException
augmentIOError ioe@IOError{ ioe_filename = fp } fun h
  = ioe { ioe_handle = Just h, ioe_location = fun, ioe_filename = filepath }
  where filepath
          | Just _ <- fp = fp
          | otherwise = case h of
                          FileHandle path _     -> Just path
                          DuplexHandle path _ _ -> Just path

hGetContents' :: Handle -> IO String
hGetContents' handle = do
    es <- wantReadableHandle "hGetContents'" handle (strictRead handle)
    case es of
      Right s -> return s
      Left e ->
          case fromException e of
            Just ioe -> throwIO (augmentIOError ioe "hGetContents'" handle)
            Nothing -> throwIO e
```

And further down:

```hs
mkHandleFromFDNoFinalizer
   :: FD.FD
   -> IODeviceType
   -> FilePath  -- a string describing this file descriptor (e.g. the filename)
   -> IOMode
   -> Bool      --  *set* non-blocking mode on the FD
   -> Maybe TextEncoding
   -> IO Handle

mkHandleFromFDNoFinalizer fd0 fd_type filepath iomode set_non_blocking mb_codec
  = do
#if !defined(mingw32_HOST_OS)
    -- turn on non-blocking mode
    fd <- if set_non_blocking
             then FD.setNonBlockingMode fd0 True
             else return fd0
#else
    let _ = set_non_blocking -- warning suppression
    fd <- return fd0
#endif

    let nl | isJust mb_codec = nativeNewlineMode
           | otherwise       = noNewlineTranslation

    case fd_type of
        Directory ->
           ioException (IOError Nothing InappropriateType "openFile"
                           "is a directory" Nothing Nothing)

        Stream
           -- only *Streams* can be DuplexHandles.  Other read/write
           -- Handles must share a buffer.
           | ReadWriteMode <- iomode ->
                mkDuplexHandleNoFinalizer fd filepath mb_codec nl

        _other ->
           mkFileHandleNoFinalizer fd filepath iomode mb_codec nl
```

Reply:
`augmentIOError` looks weird to me: there is `ioe_handle` to store a handle if one wishes so. `instance Exception IOException` makes an effort to print both `ioe_handle` and `ioe_filename`, so no information will be lost.

Reply:
From what I understand is that `FileHandle fp mvar` (Handle type) is abused somewhere to store a non-filepath.

I haven't looked deeper yet.

Reply:
The corruption happens in 
```haskell
fdToHandle :: Posix.FD -> IO Handle
fdToHandle fdint = do
   iomode <- Posix.fdGetMode fdint
   (fd,fd_type) <- FD.mkFD fdint iomode Nothing
            False{-is_socket-}
              -- NB. the is_socket flag is False, meaning that:
              --  on Windows we're guessing this is not a socket (XXX)
            False{-is_nonblock-}
              -- file descriptors that we get from external sources are
              -- not put into non-blocking mode, because that would affect
              -- other users of the file descriptor
   let fd_str = "<file descriptor: " ++ show fd ++ ">"
   mkHandleFromFD fd fd_type fd_str iomode False{-non-block-}
                  Nothing -- bin mode
```

Reply:
`file-io` uses `fdToHandle` in `openFile` / `openExistingFile` and `fdToHandle'` in Windows-specific code. I think we should set `Handle` fields properly, after retrieving it from `fdToHandle`.